### PR TITLE
IComponentContainer and Heuristic tuning

### PR DESCRIPTION
**Pull request by hopshackle**
This contains two distinct pieces of work:

1) Implementation of Proposal #133 to standardise deck visibility across everything and make tracking levels of hidden information much easier; without requiring game implementers to do too much additional work.

2) OSLAHeuristic to enable quick and easy tuning of Heuristic agents (it's already possible to tune the Heuristic used in RMHC or MCTS, so this is filling in a gap).

### CODE REVIEW

File: src/main/java/games/coltexpress/ColtExpressForwardModel.java [hunk 1]
```diff
@@ -60,12 +61,11 @@ public void _setup(AbstractGameState firstState) {
...
             playerCards.shuffle(new Random(cep.getRandomSeed()+playerIndex));
 
-            Deck<ColtExpressCard> playerHand = new Deck<>(
-                    "playerHand" + playerIndex, playerIndex);
+            Deck<ColtExpressCard> playerHand = new Deck<>("playerHand" + playerIndex, playerIndex, VisibilityMode.VISIBLE_TO_OWNER);
 
             cegs.playerHandCards.add(playerHand);
 
-            Deck<Loot> loot = new Deck<>("playerLoot" + playerIndex, playerIndex);
+            Deck<Loot> loot = new Deck<>("playerLoot" + playerIndex, playerIndex, VisibilityMode.VISIBLE_TO_OWNER);
```
Comment (rdgain) [inline src/main/java/games/coltexpress/ColtExpressForwardModel.java:68]: This is actually not visible to the owner either, hidden to all. No one knows what loot values they collected until the end of the game. Although I see later that this means to distinguish between types of loot while loot values should be kept hidden. So I guess this doesn't change much the logic?

File: src/main/java/core/CoreConstants.java [hunk 2]
```diff
@@ -34,4 +34,15 @@
...
+
+    /**
+     * Used in Components that contain other Components (see IComponentContainer) to mark which players can see the
+     * contents.
+     * MIXED_VISIBILITY is an indicator that none of the previous three apply, and that the IComponentContainer
+     * will need to implement more sophisticated logic. This is done for example in PartialObservableDeck - and
+     * this should cover almost all future eventualities.
+     */
+    public enum VisibilityMode {
+        VISIBLE_TO_ALL, HIDDEN_TO_ALL, VISIBLE_TO_OWNER, MIXED_VISIBILITY
```
Comment (rdgain) [inline src/main/java/core/CoreConstants.java:46]: Some others that I'd expect to see often would be FIRST_VISIBLE (to all) or LAST_VISIBLE (to all), for the cases where you have a deck on the table that's technically face up, so you can see the first card, but you can't look through it, so the others are technically hidden. Then if a card is removed from the deck the next one on the deck would become the one visible. An immediate example of use would be in Colt Express for the round cards where visibility of the first card is set manually, but I imagine more could use this as a default implemented functionality.

**Review by rdgain — APPROVED** [review]
Very nice! I think things look much cleaner with code taken off the game implementations as well and hidden behind intuitive to use visibility modes. The component container also makes things clearer to understand all the stacking that's going on in a game, thanks!

File: src/main/java/games/coltexpress/ColtExpressForwardModel.java (hunk 1 above)
Comment (hopshackle) [reply src/main/java/games/coltexpress/ColtExpressForwardModel.java:68]: Oh! Whenever we play Colt Express we always look at the loot tiles we pick up! Ooops. (which is why I used VISIBLE_TO_OWNER)

But...the main substantive point is that you do know the important stuff (just not which loot bags are 250 versus 500). Actually, thinking about this, this should probably be VISIBLE_TO_ALL on this logic.

File: src/main/java/core/CoreConstants.java (hunk 2 above)
Comment (hopshackle) [reply src/main/java/core/CoreConstants.java:46]: Nice idea!

**Pull request merged**